### DISCUSSION TITLE
Not able to inherit from WiFiInterface and NetworkStack classes, not able to properly set Include Paths

### DISCUSSION THREAD

**Discussion by Aditya-A-garwal** [mbed-ce/mbed-os#195] [Q&A] [answered]
Hello, I am new to MBed OS and I am trying to create my own WiFiInterface using an ESP32 as a programming exercise. To start out, I have decided to add all the code in my ```main.cpp``` file itself rather than create multiple header and sources files in a separate directory. This might seem like a silly question, but I am not able to include the ```NetworkStack.h``` and ```WiFiInterface.h``` files into my program (the compiler can not find the files). I am also not able to inherit from the ```NetworkStack``` and ```WiFiInterface``` classes without including these files. Can someone please help me out?

My mbed_app.json5 file -
```json
{
	"target_overrides": {
		"*": {
			"platform.stdio-baud-rate": 115200,
			"platform.stdio-buffered-serial": 1,
		}
	}
}
```

My CMakeLists.txt -
```cmake
cmake_minimum_required(VERSION 3.19)
cmake_policy(VERSION 3.19)

set(CMAKE_EXPORT_COMPILE_COMMANDS ON)

# Initialize Mbed OS build system.
# Note: This block must be before the include of app.cmake
set(MBED_APP_JSON_PATH mbed_app.json5)

include(mbed-os/tools/cmake/app.cmake)
add_subdirectory(mbed-os)

project(wifi-driver)

# add subdirectories and build targets here
add_executable(wifi-driver main.cpp)
target_link_libraries(wifi-driver mbed-os)
mbed_set_post_build(wifi-driver)

mbed_finalize_build()
```

Other than this, I have followed the new project setup guide from the Wiki and have made sure that everything else works fine. I am suspecting the problem is in the mbed_app.json5 file, which I have not been able to correctly populate for my purpose due to insufficient documentation, but am not sure about the issue. Any help is appreciated.

**Comment by JohnK1987** [comment]
Hello and welcome to MbedCE.

the issue will be probably in cmake settings but also depends on your MCU target. I am not so much familiar with Cmake like Jamie but I believe he will fill correct information ASAP.
Until that I thing the issue is will be on this line of CmakeLists.txt

 > target_link_libraries(wifi-driver mbed-os)

will be necessary to fill more libraries here I think. Probably something like

> target_link_libraries(wifi-driver mbed-os mbed-wifi mbed-netsocket)

BTW there were already ESP32 driver
https://github.com/ARMmbed/esp32-driver

**Reply by Aditya-A-garwal** [reply]
Hey John
I'm aware of the existing ESP32 driver, but I want to write my own without using it as a learning exercise. The issue I am facing is that I am not able to **include the header files** into my code without some ad-hoc changes to the ```CMakeLists.txt``` file. I am having to add the following line to my ```CMakelists.txt``` file
```cmake
target_include_directories(wifi-driver PUBLIC
    mbed-os/connectivity/netsocket/include
)
```

The library is automatically being linked without adding any extra code. Is this the correct way to set the include path according to the MBed CE standard? If not then what is the correct way to do so.

**Reply by JohnK1987** [reply, marked as answer]
Ok, I understand and respect your decision about old driver.

I probably do not understand your issue properly. You wrote about a code in main.cpp where you need NetworkStack.h and WiFiInterface.h

Basic MbedOS
![image](https://github.com/mbed-ce/mbed-os/assets/23274499/8fc84bff-94d4-4851-8983-bf9131d29481)
Modified top level CMakelists.txt
![image](https://github.com/mbed-ce/mbed-os/assets/23274499/833017f9-0a02-4abf-a178-a215e3f95048)
BR, Jan

**Reply by Aditya-A-garwal** [reply]
I just saw where I was going wrong. What you are saying is absolutely right and it fixed my issue. Thanks a lot John!

**Reply by JohnK1987** [reply]
I'm glad I could help. 
Good luck with your personal project and will be nice when you share your result with us.

BR, Jan
